### ISSUE TITLE
Any chance for SFTP (SSH) upload?

### ISSUE THREAD

**Issue by Tony-il-Capo** [mendhak/gpslogger#523]
Hi there
Would you consider to add a simple sftp client in the app so that we could use it to upload instead of the insecure FTP(s)? 
Maybe too much work? I think it would be beneficial, specially over normal FTP, that offers pretty much no security and should not be used at all.

**Comment by bill-mcgonigle** [comment]
@Tony-il-Capo : just FYI I'm using FolderSync Lite as a workaround, with the log files set to change their name based on the day for data consistency.  It's not ideal, granted.

**Comment by Tony-il-Capo** [comment]
@bill-mcgonigle 
3rd party and non open-source, lots of privacy issues there.

Btw I use SSH tunneling and port forwarding to log to my own web server, server is simple, stores only last values and runs in LAN only, then people at home can see where I am. It's not ideal either because SSH tunnel needs to be open (Connectbot app working), but it's very secure.

**Comment by mendhak** [comment]
I do like SSH over everything else but yes as you said, it's the amount of work, so that's why I also recommend other apps like FolderSync to do the work as they'll be better at uploading to whatever's not in GPSLogger (including SFTP).  Off the top of my head, whenever I think about SSH/SCP/SFTP I go down a rabbit hole in terms of what will be needed - key management, keys with passphrases, accepting/change/revoke host keys.  It's a world unto itself, and there are always the extra quirks/features/special-uses that get tacked on to it.

I'll leave this open, it would be a good addition to the list of upload-destinations.

**Comment by Tony-il-Capo** [comment]
Yes, it may take some work, but would probably be the most secure form of uploading GPSLogger data to private servers. Great for privacy concerned users, especially if this would be implemented  in a 'light version' of GPSLogger, as I already suggested early.

**Comment by bill-mcgonigle** [comment]
Is there an open source SFTP library for Android?  @mendhak is right about the complexity; it would be great to incorporate an existing project where all of that is properly solved already (writing perfect secure comms is not a trivial task - reuse of common code is much better when feasible).

**Comment by mendhak** [comment]
The main ones I've found:

* [JSch](http://www.jcraft.com/jsch/examples/) which does have examples for Scp/Sftp uploads
* [SSHJ](https://github.com/hierynomus/sshj) 
* [Jaramiko](https://github.com/freznicek/jaramiko) a port of paramiko but last updated 6 years ago
* [Apache Commons VFS](https://commons.apache.org/proper/commons-vfs/) 

So far it seems SSHJ and Apache Commons VFS seem to be the ones to use.  Apache VFS uses JSch internally.

**Comment by bill-mcgonigle** [comment]
My $0.02: Apache Commons VFS could enable some neat future features - cifs would be appreciated by some users, WebDAV is a tiny step away from CalDAV support, etc.

**Comment by mendhak** [comment]
Well this was frustrating.  Apache VFS seems to be built mainly for [Eclipse Equinox](http://www.eclipse.org/equinox/), and seems to have a lot of implicit dependencies which aren't well documented.  It also needed things from an Apache Commons Sandbox but I've been going in circles trying to trace various errors.  But from what I see, it seems to work just fine in that one particular dev environment, all others are a your-mileage-may-vary situation.

Will try SSHJ or JSCH later.

**Comment by mendhak** [comment]
At last I've implemented SFTP auto-sender.  I have an APK ready if anyone wants to help with testing. 

[APK download](https://drive.google.com/open?id=0B6IOK82n4BkAazlVd2JMZTB3ZmM), [PGP sign](https://drive.google.com/open?id=0B6IOK82n4BkAbHN3b2pvaE5jUHM), [SHA256 checksum](https://drive.google.com/open?id=0B6IOK82n4BkAWTlQWmFyWDZLc3M)

Some additional notes - 

I used JSch library.  It's not that straightforward but not too bad either. It does make a lot of assumptions about your app, assuming it's a console app or Swing app.

I've implemented host key checking.  It's frustrating that most examples online have host key checking disabled.  I've done it so when you do a test upload, you should get the fingerprint shown to you.  After accepting it, subsequent uploads should work. 

Had a look at doing SCP uploads as well but its code is much larger and more complex, so I'm leaving it alone.  Basically this means, special servers which use `rssh` or `scponly` won't work.

**Comment by Tony-il-Capo** [comment]
I will definitely check/test that. I will keep you posted about eventual problems.

**Comment by Tony-il-Capo** [comment]
Hi @mendhak 
I've successfully uploaded gps data to my private SSH server, but I have some questions:

- How is the passphrase stored? I mean, is it somehow protected or stored somewhere in plain text? This is critical for private servers, we certainly don't want others to have it in case the phone gets lost, all private LAN could be exposed...

- Could be an option to keep in memory the passphrase for the session only? It would ask for it at the start of logging and we would need to type it for every session, like a regular SSH client.
____
On a side note I think you should also clarify the Privacy Policy about third-partys. I think that is clear that GPSLogger does not collect and send any personal data to the 'GPSLogger organization' (developers), **but does it collect and send any data to third-party services** (i.e analytics, tracking and profiling)**?**

**Comment by mendhak** [comment]
The SFTP values (user, password, passphrase) are stored in the SharedPreferences which is a secured file individual to each application.  [Read more](http://stackoverflow.com/questions/3608883/how-secure-are-sqlite-and-sharedpreferences-files-on-android).  So it gets protected against other applications, but it won't protect against the user themselves. If a user roots their phone, they can see it all.  If a user's phone is stolen, same applies - unless you [encrypt your file system](https://source.android.com/security/encryption/#full-disk) supported in Android 5+.  In those scenarios when a phone is stolen you need to wipe it via Android device manager.

Next is the private LAN being exposed - this is usually where chrooting comes in, or individual keys for specific purposes; SSH admins will create individual users + keypairs for the specific purpose of uploads from a single app and that's what the app uses.  In other words I am saying that the access-privileges of the system being accessed by any SSH app is not the scope of the app itself, it's down to the admin.  

As for a session-only option, please see my very first comment in this thread.  It's easy to go down a rabbit hole in terms of this feature-wise, but it's a huge drain on time to maintain,  I'd like to keep things as simple and straightforward as possible.  - I'll mention anyway that prompting for a user for typed passphrase will go into the same place as it does currently, and a session can last for days, so the same lost-phone scenario applies.

**Comment by mendhak** [comment]
OK Privacy policy updated in the [jsch branch](https://github.com/mendhak/gpslogger/blob/jsch/assets/text/privacypolicy.md) to say nothing sent to third parties

**Comment by mendhak** [comment]
Updated APK.  Now I've added a "clear authorization" button which makes the remote server untrusted again. Also added an SFTP icon.  I can't find a good one, I think this should do though...

APK: https://drive.google.com/open?id=0B6IOK82n4BkAazlVd2JMZTB3ZmM
PSP Signature: https://drive.google.com/open?id=0B6IOK82n4BkAbHN3b2pvaE5jUHM
SHA 256 Checksum: https://drive.google.com/open?id=0B6IOK82n4BkAWTlQWmFyWDZLc3M
